### ISSUE TITLE
confused about sess.run() function

### ISSUE THREAD

**Issue by helenahan2015** [aymericdamien/TensorFlow-Examples#22]
I'm trying to understand this code https://github.com/aymericdamien/TensorFlow-Examples/blob/master/examples/5%20-%20User%20Interface/loss_visualization.py. There are 3 sess.run() functions in the "for" loop (see attached picture). Does this mean the whole graph runs **3** times in **one** "for" loop? 
![capture](https://cloud.githubusercontent.com/assets/16149289/14101499/7d3efe3c-f5c7-11e5-80de-9f10cb79a232.PNG)
Could I use only one "sess.run()" instead? (I tried to use _result = sess.run(optimizer, cost, merged_summary_op)_, but it seems the result is incorrect)

**Comment by aymericdamien** [comment]
It doesn't really run the 'whole' graph 3 times for 1 loop, it only runs the computation subgraph necessary for every operation you run (So calling `cost` will not compute gradients like `optimizer` op does). But considering calling `run` need to initialize some parameters, `result = sess.run([optimizer, cost, merged_summary_op])` should be more efficient (as it only calls run once vs 3 before). But it should be a big difference.
And about only using sess.run([cost,merged_summary_op]), do you run `optimizer` op before or not?

**Comment by helenahan2015** [comment]
@aymericdamien Thank you for your explanation. For your last question, I only use sess.run([cost,merged_summary_op]) and there is no sess.run(optimizer) before or after. And I wonder how to run the whole graph?

**Comment by aymericdamien** [comment]
I think this should works... Can you try to just run sess.run([cost])?

**Comment by helenahan2015** [comment]
@aymericdamien I tried just run sess.run([cost]), here is code and result, obviously, the result is incorrect. 
![capture1](https://cloud.githubusercontent.com/assets/16149289/14196673/6fb6487c-f7f9-11e5-9325-9705e075598b.PNG)
![capture2](https://cloud.githubusercontent.com/assets/16149289/14196676/736f7eca-f7f9-11e5-98e9-d0d13df1e854.PNG)

I think it makes sense and it shouldn't work. As you said, sess.run([cost]) only runs the subgraph. So only cost op is calculated.
![capture](https://cloud.githubusercontent.com/assets/16149289/14196773/1bf8c920-f7fa-11e5-9574-ab54ba7876d7.PNG)

**Comment by aymericdamien** [comment]
Results are correct, because you haven't started to train (by running the optimizer op) your network has random weights, so it give you such results. You can see accuracy is ~0.1, which is same as a random guess out of 10 classes of MNIST.
You have to use optimizer to train your model (by reducing the cost, applying backpropagation algorithm and updating variable weights).

**Comment by Xikhari** [comment]
I am late, but yes. When you run a session you just send the parameters to the function that fetches and

feed_dict just load the parameters in the placeholders that the function you are running will use. I find it easier to see at it as a C function receiving a pointer. So basically it would be like a function witch takes a function and a structure and later it makes a pointer to the data and passes it to the function provided. Not really intuitive and pythonic but that is tf.

**Comment by jmgrn61** [comment]
I think it would not work, if we only do sess.run([optimizer]), and "cost" has to be there.
For example,  sess.run([optimizer, cost]).
I am really confused because "cost" is already included in "optimizer", why can not we just use "optimizer"?

**Comment by VXU1230** [comment]
@jmgrn61  I'm in the same boat. Confused why we put optimizer before cost when running the session.

**Comment by ParichayDidwania** [comment]
@jmgrn61 In case if you are using an Iterator then calling sess.run(cost) will give u the next batch of images and labels from the iterator. Same occurs if you call sess.run(optimizer). Thus, i think that if you call these two separately then its actually showing the cost of batch 'n' and optimizing based on the batch 'n+1' therefore completely skipping the 'nth' batch. So i always use them in same sess.run().

I also tested this and found out that iterations reduced when using seperately as 1 batch was always getting skipped in each Loop.

**Comment by ParichayDidwania** [comment]
@jmgrn61 In case if you are using an Iterator then calling sess.run(cost) will give u the next batch of images and labels from the iterator. Same occurs if you call sess.run(optimizer). Thus, i think that if you call these two separately then its actually showing the cost of batch 'n' and optimizing based on the batch 'n+1' therefore completely skipping the 'nth' batch. So i always use them in same sess.run().

![Capture](https://user-images.githubusercontent.com/47176112/58349902-29952800-7e82-11e9-9428-1da0e20f188b.JPG)

I also tested this and found out that iterations reduced when using them in separate sess.run() as 2 batches were always getting skipped in each Loop (1st skip on sess.run(loss), 2nd skip on sess.run(accuracy). Only set of images which is being trained is at sess.run(optimizer). Thus i am skipping every 2 batches per loop, meaning i am only training on 1/3 of my total training dataset.
